Unboxed kernels in c10

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23695 [wip] Fixed size arrays
* #23668 [wip] Call aten ops through c10 dispatcher
* #23888 Allow TensorMethods.h to include Dispatcher.h (alternative)
* #23667 Register ATen ops with c10
* #23748 Disambiguate tensor and string ops
* #23694 toString(FunctionSchema) shows overload name
* #23666 c10 dispatcher stores autograd kernels
* #23665 Allow kernels that don't have a boxed version
* **#23447 Unboxed kernels in c10**
* #23834 Open up AliasAnalysisKind for any ops

The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Unboxed kernels in c10

The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

dzhulgakov commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -247,6 +236,30 @@ class DispatchTable final {
+      [&] (const detail::KernelTable_& table) -> const DispatchTableEntry& {
+        // We have a dispatch table. Find the correct kernel for the inputs and return it.
+
+        TORCH_INTERNAL_ASSERT(dispatch_strategy_.is_valid_, "Operator ", operator_name_, " has an invalid dispatch key but kernels registered.");
```
seems like a duplicated check

smessmer replied:
duplicate to where?

dzhulgakov commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -247,6 +236,30 @@ class DispatchTable final {
+
+        TORCH_INTERNAL_ASSERT(dispatch_strategy_.is_valid_, "Operator ", operator_name_, " has an invalid dispatch key but kernels registered.");
+
+        TensorTypeId dispatch_key = getDispatchKey();
```
it's probably not a comment for now, but this code is pretty convoluted just to optimize the catch-all case, right? given that majority (all?) of our kernels would be using dispatch, we could have made this code more readable by folding all lambdas and just passing in the dispatch key directly.

smessmer replied:
most custom ops don't use dispatch and internally, meta ops don't either.

dzhulgakov commented on aten/src/ATen/core/op_registration/test_helpers.h:
```diff
@@ -34,6 +34,12 @@ inline std::vector<c10::IValue> callOp(const c10::OperatorHandle& op, Args... ar
 }
 
+template<class Result, class... Args>
+inline Result callOpUnboxed(const c10::OperatorHandle& op, c10::TensorTypeId dispatchKey, Args... args) {
```
I though the whole point of unboxed call was to figure out the dispatch key from arguments. Do you think it's doable with template magic to collect keys of arguemnts (I recall you had this code somewhere). If not - we can codegen it

smessmer replied:
This is planned but a future PR. There's a mismatch between c10 and aten currently because c10 chooses "the first tensor argument" to dispatch on while aten looks for "the tensor argument named 'self'". The plan is to add this 'self' functionality to c10 and then use it.

dzhulgakov commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -159,26 +160,14 @@ class DispatchTable final {
-                  "'. Tried to look up kernel for dispatch key '", toString(dispatch_key),
-                  "'. Registered dispatch keys are: ", listAllDispatchKeys());
+     return lookup_([=] {
+       TORCH_INTERNAL_ASSERT(dispatch_strategy_.is_valid_, "Operator ", operator_name_, " has an invalid dispatch key but kernels registered.");
```
duplicated with here

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

dzhulgakov reviewed (approved):
Stamping it as a standalone change. But it'd be nice to see how it'd connect end-to-end (including boxing/unboxing and figuring out the dispatch keys) and what's the plan going forward

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)

smessmer committed:
Update on "Unboxed kernels in c10"


The c10 dispatcher now also stores a `void*` pointer to the unboxed kernel function and this kernel function can be called if the call site knows the exact kernel signature.

It is not clear if this API will survive in the long term, but in the short term this allows an easier migration from ATen to c10 and is supposed to replace ATenDispatch.

Differential Revision: [D16521939](https://our.internmc.facebook.com/intern/diff/D16521939/)
